controller.go

```diff
@@ -344,15 +299,118 @@ func (c *controller) handleCRDeletion(obj interface{}) {
 		if err := c.notifyClient.Send(note); err != nil {
 			Logger.Error(err, "Failed to send notification", "namespace", ns)
 		}
-		c.stateHistory.Lock()
-		c.stateHistory.perNamespaceHistory[ns] = &NamespaceStateHistory{
-			mongoStatus:  mongoStatus,
-			backupStatus: backupStatus,
-			notification: notificationstate,
-			lastUpdate:   time.Now(),
+	}
+	Logger.Info("Skipping Notification. Curent Status: ", "NameSpace", ns, "Backup", backupStatus, "Mongo", mongoStatus, "Notification", notificationstate, "Event", "CR Deletion")
+}
+
+func (c *controller) handleBackupAndMongoCreateUpdateEvents() bool {
+	var mongoStatus, backupStatus, schedulerStatus Status
+	item, quit := c.backupqueue.Get()
+	if quit {
+		return false
+	}
+	defer c.backupqueue.Done(item)
+
+	u := item.(*unstructured.Unstructured)
+	ns := u.GetNamespace()
+	creationTime := u.GetCreationTimestamp()
+
+	backupStatus = getCRStatus(c.backupLister, ns)
+	mongoStatus = getCRStatus(c.mongoLister, ns)
+
+	state := notification.StatesAndNotificationsMapping[string(backupStatus)][string(mongoStatus)]
+
+	if ((backupStatus == NOTFOUND && mongoStatus != NOTFOUND) ||
+		(backupStatus != NOTFOUND && mongoStatus == NOTFOUND)) &&
+		(time.Since(creationTime.Time) < time.Duration(1*time.Minute)) {
+		// It might happen that Mongo Cr is created and backup is not created yet or vice versa,
+		// In this case we dont want to send unreachable or deleted straightway
+		// following e.g. scenarios
+		// Mongo --> Pending/Available/NotReachable   AND   Backup --> NotFound
+		// then we send Provisioning till creationTime < 2 min
+		// same goes for below case
+		// Backup --> Pending/Available/ and Mongo --> NotFound
+		// Idea here is to wait for 2 minutes before we send notification as defined in StatesAndNotificationsMapping \
+		// because CR creation might be delayed or cache has not been sync properly
+		state = "Provisioning"
+	}
+
+	if state == "Success" {
+		schedulerStatus = getSchedulerStatus(creationTime, ns)
+		Logger.Info("", "SchedulerStatus", schedulerStatus, "NameSpace", ns)
+		state = notification.BackupAndSchedulerStatusMapping[string(AVAILABLE)][string(schedulerStatus)]
+		if state == "Provisioning" {
+			defer c.backupqueue.AddAfter(item, time.Duration(RetryDelaySeconds)*time.Second)
+		} else {
+			defer c.backupqueue.Forget(item)
+		}
+	}
+
+	if c.skipNotification(ns, state, backupStatus, mongoStatus) {
+		return true
+	}
+
+	note := notification.Note{
+		State:          state, //TODO: handle unknown state transition error
+		Namespace:      ns,
+		FailureMessage: "",
+	}
+
+	err := c.notifyClient.Send(note) //TODO: check if notification send failed and retry in case of non 200
+	if err != nil {
+		Logger.Error(err, "Failed to send notification")
+	}
+
+	return true
+}
+
+func getSchedulerStatus(creationTime v1.Time, ns string) Status {
```

Comment:
you can form a Scheudle struct and put the relate methods (getgetSchedulerStatus, IsBackupSchedulerReady) under that